File: Applied Course/4.Problem Solving/1.Problems on Array/22.Set Matrix Zeros.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int r=matrix.size(),c=matrix[0].size();
        set<int> row,col;
        for(int i=0;i<r;i++)
            for(int j=0;j<c;j++)
                if(matrix[i][j]==0)
                {
                    row.insert(i);
                    col.insert(j);
                }
        for(int i=0;i<r;i++)
            for(int j=0;j<c;j++)
                if(row.find(i)!=row.end() || col.find(j)!=col.end())
                    matrix[i][j]=0;
    }
};
// Logic-2
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        bool rowFlag=false,colFlag=false;
        int row=matrix.size(),col=matrix[0].size();
        for(int i=0;i<row;i++)
        {
            for(int j=0;j<col;j++)
            {
                if(i==0 && matrix[i][j]==0)
                    rowFlag=true;
                if(j==0 && matrix[i][j]==0)
                    colFlag=true;
                if(matrix[i][j]==0)
                {
                    matrix[0][j]=0;
                    matrix[i][0]=0;
                }
            }
        }
        for(int i=1;i<row;i++)
            for(int j=1;j<col;j++)
                if(matrix[i][0]==0 || matrix[0][j]==0)
                    matrix[i][j]=0;
        if(rowFlag==true)
            for(int i=0;i<col;i++)
                matrix[0][i]=0;
        if(colFlag==true)
            for(int j=0;j<row;j++)
                matrix[j][0]=0;
    }
};
int main()
{
    // https://leetcode.com/problems/set-matrix-zeroes/
    return 0;
}

File: Applied Course/4.Problem Solving/1.Problems on Array/22.Set Matrix Zeros.py
# https://leetcode.com/problems/set-matrix-zeroes/
# Logic-1
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])
        rows = set()
        columns = set()
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    rows.add(i)
                    columns.add(j)
        for i in range(m):
            for j in range(n):
                if i in rows or j in columns:
                    matrix[i][j] =0
        return matrix
# Logic-2
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        row_flag, col_flag = False, False
        m = len(matrix)
        n = len(matrix[0])
        #Traverse the 
        for i in range(m):
            for j in range(n):
                if i==0 and matrix[i][j] == 0:
                    row_flag = True
                if j == 0 and matrix[i][j] == 0:
                    col_flag = True
                if matrix[i][j] == 0:
                    matrix[0][j] = 0
                    matrix[i][0] = 0
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0
        if row_flag == True:
            for i in range(n):
                matrix[0][i] = 0
        if col_flag:
            for j in range(m):
                matrix[j][0] = 0
        return matrix